File: data/src/test/java/com/github/anrimian/musicplayer/data/controllers/music/players/CompositeMediaPlayerTest.java
package com.github.anrimian.musicplayer.data.controllers.music.players;

import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource;
import com.github.anrimian.musicplayer.domain.models.player.error.ErrorType;
import com.github.anrimian.musicplayer.domain.models.player.events.ErrorEvent;
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent;

import org.junit.Before;
import org.junit.Test;
import org.mockito.InOrder;
import org.mockito.Mockito;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.observers.TestObserver;
import io.reactivex.rxjava3.subjects.PublishSubject;
import utils.TestDataProvider;

import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class CompositeMediaPlayerTest {

    private final AppMediaPlayer player1 = mock(AppMediaPlayer.class);
    private final AppMediaPlayer player2 = mock(AppMediaPlayer.class);

    private CompositeMediaPlayer compositeMediaPlayer;

    private InOrder inOrder = Mockito.inOrder(player1, player2);

    private final PublishSubject<PlayerEvent> player1EventSubject = PublishSubject.create();
    private final PublishSubject<PlayerEvent> player2EventSubject = PublishSubject.create();

    private final PublishSubject<Long> player1PositionSubject = PublishSubject.create();
    private final PublishSubject<Long> player2PositionSubject = PublishSubject.create();

    @Before
    public void setUp() {
        when(player1.getTrackPositionObservable()).thenReturn(player1PositionSubject);
        when(player1.getEventsObservable()).thenReturn(player1EventSubject);
        when(player1.getSpeedChangeAvailableObservable()).thenReturn(Observable.just(true));


        when(player2.getTrackPositionObservable()).thenReturn(player2PositionSubject);
        when(player2.getEventsObservable()).thenReturn(player2EventSubject);
        when(player2.getSpeedChangeAvailableObservable()).thenReturn(Observable.just(true));

        compositeMediaPlayer = new CompositeMediaPlayer(() -> player1, () -> player2);
    }

    @Test
    public void testPlayersSwitch() {
        CompositionSource composition = TestDataProvider.fakeCompositionSource(0);

        compositeMediaPlayer.prepareToPlay(composition, 0L, null);
        inOrder.verify(player1).prepareToPlay(eq(composition), eq(0L),  eq(null));

        player1EventSubject.onNext(new ErrorEvent(ErrorType.UNSUPPORTED, composition));
        inOrder.verify(player1).release();
        inOrder.verify(player2).prepareToPlay(eq(composition), eq(0L),  eq(ErrorType.UNSUPPORTED));
    }

    @Test
    public void testAllPlayersNotWorking() {
        CompositionSource composition = TestDataProvider.fakeCompositionSource(0);

        TestObserver<PlayerEvent> eventsObserver = compositeMediaPlayer.getEventsObservable().test();

        compositeMediaPlayer.prepareToPlay(composition, 0L, null);
        inOrder.verify(player1).prepareToPlay(eq(composition), eq(0L),  eq(null));

        player1EventSubject.onNext(new ErrorEvent(ErrorType.UNSUPPORTED, composition));
        inOrder.verify(player1).release();
        inOrder.verify(player2).prepareToPlay(eq(composition), eq(0L),  eq(ErrorType.UNSUPPORTED));

        player2EventSubject.onNext(new ErrorEvent(ErrorType.UNSUPPORTED, composition));

        eventsObserver.assertValue(new ErrorEvent(ErrorType.UNSUPPORTED, composition));

        CompositionSource composition2 = TestDataProvider.fakeCompositionSource(2);

        compositeMediaPlayer.prepareToPlay(composition2, 0L, null);
        inOrder.verify(player2).release();
        inOrder.verify(player1).prepareToPlay(eq(composition2), eq(0L),  eq(null));
    }

    @Test
    public void testPlayersSwitchWithPosition() {
        CompositionSource composition = TestDataProvider.fakeCompositionSource(0);

        compositeMediaPlayer.prepareToPlay(composition, 0L, null);
        inOrder.verify(player1).prepareToPlay(eq(composition), eq(0L), eq(null));

        player1PositionSubject.onNext(100L);

        player1EventSubject.onNext(new ErrorEvent(ErrorType.UNSUPPORTED, composition));
        inOrder.verify(player1).release();
        inOrder.verify(player2).prepareToPlay(eq(composition), eq(100L), eq(ErrorType.UNSUPPORTED));
    }
}

File: data/src/test/java/com/github/anrimian/musicplayer/data/controllers/music/players/CompositeMediaPlayerTest.kt
package com.github.anrimian.musicplayer.data.controllers.music.players

import com.github.anrimian.musicplayer.domain.models.composition.content.CompositionContentSource
import com.github.anrimian.musicplayer.domain.models.composition.content.RelaunchSourceException
import com.github.anrimian.musicplayer.domain.models.composition.content.UnsupportedSourceException
import com.github.anrimian.musicplayer.domain.models.player.SoundBalance
import com.github.anrimian.musicplayer.domain.models.player.events.MediaPlayerEvent
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.observers.TestObserver
import io.reactivex.rxjava3.subjects.PublishSubject
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentMatchers.anyFloat
import org.mockito.kotlin.any
import org.mockito.kotlin.anyOrNull
import org.mockito.kotlin.eq
import org.mockito.kotlin.inOrder
import org.mockito.kotlin.mock
import org.mockito.kotlin.never
import org.mockito.kotlin.whenever

class CompositeMediaPlayerTest {

    private val player1: AppMediaPlayer = mock()
    private val player2: AppMediaPlayer = mock()
    
    private lateinit var compositeMediaPlayer: CompositeMediaPlayer
    
    private val inOrder = inOrder(player1, player2)
    
    private val player1ErrorEventSubject = PublishSubject.create<MediaPlayerEvent>()
    private val player2ErrorEventSubject = PublishSubject.create<MediaPlayerEvent>()
    
    private val player1PositionSubject = PublishSubject.create<Long>()
    private val player2PositionSubject = PublishSubject.create<Long>()

    private lateinit var textEventObserver: TestObserver<MediaPlayerEvent>
    
    @BeforeEach
    fun setUp() {
        whenever(player1.prepareToPlay(any(), anyOrNull())).thenReturn(Completable.complete())
        whenever(player1.getTrackPositionObservable()).thenReturn(player1PositionSubject)
        whenever(player1.getPlayerEventsObservable()).thenReturn(player1ErrorEventSubject)
        whenever(player1.getSpeedChangeAvailableObservable()).thenReturn(Observable.just(true))

        whenever(player2.prepareToPlay(any(), anyOrNull())).thenReturn(Completable.complete())
        whenever(player2.getTrackPositionObservable()).thenReturn(player2PositionSubject)
        whenever(player2.getPlayerEventsObservable()).thenReturn(player2ErrorEventSubject)
        whenever(player2.getSpeedChangeAvailableObservable()).thenReturn(Observable.just(true))

        compositeMediaPlayer = CompositeMediaPlayer(
            arrayListOf({ player1 }, { player2 }), 
            1f,
            SoundBalance(1f, 1f),
            false
        )
        textEventObserver = compositeMediaPlayer.getPlayerEventsObservable().test()
    }

    @Test
    fun `test players switch on prepare`() {
        val source: CompositionContentSource = mock()
        
        whenever(player1.prepareToPlay(any(), anyOrNull()))
            .thenReturn(Completable.error(UnsupportedSourceException()))
        compositeMediaPlayer.prepareToPlay(source).subscribe()
        inOrder.verify(player1).prepareToPlay(eq(source), anyOrNull())
        inOrder.verify(player1).release()
        inOrder.verify(player2).prepareToPlay(eq(source), anyOrNull())
    }

    @Test
    fun `test all players not working on prepare`() {
        val source: CompositionContentSource = mock()

        whenever(player1.prepareToPlay(any(), anyOrNull()))
            .thenReturn(Completable.error(UnsupportedSourceException()))
        val secondException = UnsupportedSourceException()
        whenever(player2.prepareToPlay(any(), anyOrNull())).thenReturn(Completable.error(secondException))
        
        compositeMediaPlayer.prepareToPlay(source)
            .test()
            .assertError(secondException)
        inOrder.verify(player1).prepareToPlay(eq(source), anyOrNull())
        inOrder.verify(player1).release()
        inOrder.verify(player2).prepareToPlay(eq(source), anyOrNull())
        inOrder.verify(player2, never()).release()
    }

    @Test
    fun `test players switch on event`() {
        val source: CompositionContentSource = mock()
        compositeMediaPlayer.prepareToPlay(source).subscribe()
        val exception = UnsupportedSourceException()
        player1ErrorEventSubject.onNext(MediaPlayerEvent.Error(exception))
        textEventObserver.assertValue { event ->
            event is MediaPlayerEvent.Error && event.throwable is RelaunchSourceException
        }

        compositeMediaPlayer.prepareToPlay(source).subscribe()

        inOrder.verify(player1).release()
        inOrder.verify(player2).setPlaybackSpeed(anyFloat())
    }

    @Test
    fun `test players switch on event to the end of available players`() {
        val source: CompositionContentSource = mock()
        val exception = UnsupportedSourceException()

        compositeMediaPlayer.prepareToPlay(source).subscribe()
        player1ErrorEventSubject.onNext(MediaPlayerEvent.Error(exception))
        textEventObserver.assertValue { event ->
            event is MediaPlayerEvent.Error && event.throwable is RelaunchSourceException
        }

        compositeMediaPlayer.prepareToPlay(source).subscribe()
        inOrder.verify(player1).release()
        inOrder.verify(player2).setPlaybackSpeed(anyFloat())

        player2ErrorEventSubject.onNext(MediaPlayerEvent.Error(exception))
        textEventObserver.assertValueAt(1) { event ->
            event is MediaPlayerEvent.Error && event.throwable is UnsupportedSourceException
        }
        compositeMediaPlayer.prepareToPlay(source).subscribe()
        inOrder.verify(player1, never()).release()
        inOrder.verify(player2, never()).setPlaybackSpeed(anyFloat())
    }

    @Test
    fun `switch player on event and prepare same source`() {
        val source: CompositionContentSource = mock()
        val exception = UnsupportedSourceException()

        compositeMediaPlayer.prepareToPlay(source).subscribe()
        player1ErrorEventSubject.onNext(MediaPlayerEvent.Error(exception))
        compositeMediaPlayer.prepareToPlay(source).subscribe()

        inOrder.verify(player1).prepareToPlay(eq(source), anyOrNull())
        inOrder.verify(player1).release()
        inOrder.verify(player2).prepareToPlay(eq(source), anyOrNull())
        inOrder.verify(player2, never()).release()
        textEventObserver.assertValue { event -> event is MediaPlayerEvent.Error }
    }

    @Test
    fun `switch player on event and prepare different source`() {
        val source: CompositionContentSource = mock()
        val source2: CompositionContentSource = mock()
        val exception = UnsupportedSourceException()

        compositeMediaPlayer.prepareToPlay(source).subscribe()
        player1ErrorEventSubject.onNext(MediaPlayerEvent.Error(exception))
        textEventObserver.assertValue { event ->
            event is MediaPlayerEvent.Error && event.throwable is RelaunchSourceException
        }

        compositeMediaPlayer.prepareToPlay(source2).subscribe()

        inOrder.verify(player1).prepareToPlay(eq(source), anyOrNull())
        inOrder.verify(player1).prepareToPlay(eq(source2), anyOrNull())
        inOrder.verify(player1, never()).release()
        inOrder.verify(player2, never()).prepareToPlay(eq(source), anyOrNull())
        inOrder.verify(player2, never()).release()
    }

}